contracts/NonFungibleToken.cdc

```diff
@@ -41,162 +36,160 @@ Collection to complete the transfer.
-        pub fun withdraw(withdrawID: UInt64): @NFT {
+        // We emit withdraw events from the provider interface because conficting withdraw
+        // events aren't as confusing to event listeners as conflicting deposit events
+
+        /// withdraw removes an NFT from the collection and moves it to the caller
+        /// It does not specify whether the ID is UUID or not
+        access(Withdrawable) fun withdraw(withdrawID: UInt64): @{NFT} {
             post {
                 result.id == withdrawID: "The ID of the withdrawn token must be the same as the requested ID"
+                emit Withdraw(type: result.getType().identifier, id: result.id, uuid: result.uuid, from: self.owner?.address, providerUUID: self.uuid)
             }
         }
     }
 
     /// Interface to mediate deposits to the Collection
     ///
-    pub resource interface Receiver {
+    access(all) resource interface Receiver {
 
-        /// Adds an NFT to the resource implementing it
-        ///
-        /// @param token: The NFT resource that will be deposited
+        /// deposit takes an NFT as an argument and adds it to the Collection
         ///
-        pub fun deposit(token: @NFT)
-    }
+        access(all) fun deposit(token: @{NFT})
 
-    /// Interface that an account would commonly 
-    /// publish for their collection
-    ///
-    pub resource interface CollectionPublic {
-        pub fun deposit(token: @NFT)
-        pub fun getIDs(): [UInt64]
-        pub fun borrowNFT(id: UInt64): &NFT
-        /// Safe way to borrow a reference to an NFT that does not panic
-        ///
-        /// @param id: The ID of the NFT that want to be borrowed
-        /// @return An optional reference to the desired NFT, will be nil if the passed id does not exist
-        ///
-        pub fun borrowNFTSafe(id: UInt64): &NFT? {
-            post {
-                result == nil || result!.id == id: "The returned reference's ID does not match the requested ID"
-            }
-            return nil
-        }
+        /// getSupportedNFTTypes returns a list of NFT types that this receiver accepts
+        access(all) view fun getSupportedNFTTypes(): {Type: Bool}
+
+        /// Returns whether or not the given type is accepted by the collection
+        /// A collection that can accept any type should just return true by default
+        access(all) view fun isSupportedNFTType(type: Type): Bool
     }
 
     /// Requirement for the concrete resource type
     /// to be declared in the implementing contract
     ///
-    pub resource Collection: Provider, Receiver, CollectionPublic {
+    access(all) resource interface Collection: Provider, Receiver {
 
-        /// Dictionary to hold the NFTs in the Collection
-        pub var ownedNFTs: @{UInt64: NFT}
+        /// Return the NFT CollectionData View
+        /// has to be AnyStruct and cast to the view later to avoid circular dependency issues
+        access(all) fun getNFTCollectionDataView(): AnyStruct
```

Comment:
Personally I do not see this as needed, we have lots of code to resolve this view other places. This is just extra code that everybody have to write.

Reply:
I would agree if I trusted everyone to implement metadata views properly, but this forces everyone to implement this function and think about how to implement the collection data view instead of just skipping or returning the bare minimum, so i think it is pretty useful

Reply:
But it returns AnyStruct? you can implement it with returning "bob". If it said you have to return NFTCollectionDataView it is entirely different.

Reply:
the point is that it forces people to think about it, which IMO is worthwhile. I'll need to update the comment on the function to include more information and maybe a link to the guide, but I want people to be forced to think about how to do it and to look for resources

Reply:
I get that, but why can it then not return NFTCollectionData? that way it is also far easier to use since you don`t have to cast and stuff.

Reply:
Agreed, but that creates a circular dependency in `NonFungibleToken` because we would have to import `MetadataViews` and we already import `NonFungibleToken` in `MetadataViews` and I don't know how to make that work. If you know a way to make that work, let me know and I can update it